### PR TITLE
Monaco test branch

### PR DESCRIPTION
**Pull request by PrateekSane**
monaco editor implimented. Properly compiles in python and html. has light/dark mode. properly saves/update state.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,17 +8,21 @@
...
     "@fortawesome/free-solid-svg-icons": "^5.15.1",
     "@fortawesome/react-fontawesome": "^0.1.12",
+    "@monaco-editor/react": "^3.6.3",
     "bootstrap": ">=4.5.3",
     "codemirror": "^5.58.2",
     "cross-env": "^7.0.0",
     "crypto-js": "^3.1.9-1",
     "firebase": "^7.24.0",
     "immutable": "^3.8.2",
+    "monaco-editor": "^0.21.2",
```
Comment (lumisphere902) [inline package.json:18]: Remove

File: package.json [hunk 2]
```diff
@@ -45,7 +49,8 @@
     "enzyme-to-json": "^3.6.1",
     "husky": "^4.3.0",
     "lint-staged": "^10.5.1",
-    "prettier": "^2.1.2"
+    "prettier": "^2.1.2",
+    "react-app-rewired": "^2.1.6"
```
Comment (lumisphere902) [inline package.json:53]: Remove

File: package.json [hunk 3]
```diff
@@ -8,17 +8,21 @@
...
     "immutable": "^3.8.2",
+    "monaco-editor": "^0.21.2",
+    
     "node-sass": "^4.14.1",
     "react": "^16.14.0",
-    "react-codemirror2": "^7.2.1",
+    
     "react-dom": "^16.14.0",
     "react-modal": "^3.11.2",
+
```
Comment (lumisphere902) [inline package.json:25]: Remove the whitespace

File: src/components/TextEditor/components/TextEditor.js [hunk 4]
```diff
@@ -200,19 +165,14 @@ class TextEditor extends React.Component {
...
-   * returns a theme string for the CodeMirror editor, based off of the app's current theme
-   * @param {string} theme - the app's current theme
-   * @returns {string} the codemirror theme - see https://codemirror.net/demo/theme.html for more info
-   */
-
+  //returns a theme based on light or dark mode for vs code name
   getCMTheme = (theme) => {
     switch (theme) {
       case "light":
         return "duotone-light";
```
Comment (lumisphere902) [inline src/components/TextEditor/components/TextEditor.js:172]: Does this need to be changed?

File: src/components/TextEditor/components/TextEditor.js [hunk 5]
```diff
@@ -312,7 +273,42 @@ class TextEditor extends React.Component {
               maxHeight: this.props.screenHeight - 61 - 20,
             }}
           >
-            <CodeMirror
+            <ControlledEditor
+              language={this.props.viewOnly ? this.props.vlanguage : this.props.language}
```
Comment (lumisphere902) [inline src/components/TextEditor/components/TextEditor.js:277]: Are you sure this is mapped properly?

File: src/components/TextEditor/components/TextEditor.js [hunk 6]
```diff
@@ -17,25 +15,12 @@ import { SketchThumbnailArray } from "../../Sketches/constants";
...
- * None
- */
-
 class TextEditor extends React.Component {
   constructor(props) {
     super(props);
 
     this.state = {
-      codeMirrorInstance: null,
+      editorInstance: null,
```
Comment (lumisphere902) [inline src/components/TextEditor/components/TextEditor.js:23]: purpose of this parameter?

Comment (PrateekSane) [reply src/components/TextEditor/components/TextEditor.js:23]: Ngl i have no idea. it was in the code mirror version. once it mounts, it says it has been made

**Comment by krashanoff** [comment]
Thanks for all the work on getting this rolling Prateek, and for taking the review Timothy. I'll leave this one to you for the moment, but feel free to @ me if anything needs more input. Reminder for code review practice is to batch your comments by clicking "Add to review" or "Start a review" rather than adding single review comments.

image

File: src/components/TextEditor/components/TextEditor.js [hunk 7]
```diff
@@ -312,7 +274,42 @@ class TextEditor extends React.Component {
               maxHeight: this.props.screenHeight - 61 - 20,
             }}
           >
-            <CodeMirror
+            <ControlledEditor
```
Comment (krashanoff) [inline src/components/TextEditor/components/TextEditor.js:277]: Nothing to review here, just wanted to comment on how much cleaner this is.

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey Prateek, thanks for your work on this! Here's my quick once-over of the changes. With an outline of the high-level functionality I think is lacking a bit. Since this is a significant feature, I think it may be best if we implement it incrementally onto a single branch off of master before merging.

## Core Functionality Review
- [x] Edit sketches
- [x] Save sketches
- [x] Swap between sketches
- [x] Fully-fledged Monaco

Really impressed with how you handled this all. Great work Prateek. One thing that I noticed needs handling is:
- [ ] View-only works properly

This is a quick fix though: disable editing of the `value` prop when you are in the viewOnly mode.

## Resizing Trouble

When the browser is resized, Monaco fails to update its dimensions, leading to incidents like this:

image

This is pretty much the only issue that interferes with the usability of the editor. Everything else is looking nice.

## Monaco Tuning

Something that encompasses a decent chunk of the issues I'm about to describe is "extra monaco features we don't really need". In a similar vein, it'd be nice to tune the editor just a bit to fit our clientele. This is really easy to handle - it'll just be a matter of configuring the editor.

* Minimap: the minimap is great for large projects, but it doesn't fit with our bite-sized kid-friendly projects.
* Internal scrollbars: having internal scrollbars within the editor is not what we want. We can disable this internally.
* Command palette simplification: right now, users can access the full command palette. We could trim it up a little bit to childproof the editor.
* Color scheme: it'd be nice if we ported our color scheme to the editor as a theme.
* Font-related
    * Font size is a bit too small if you ask me.
    * Font family should match the rest of our stack.

## Misc.
* Highlighting/suggestions for React/JSX would be nice. This would just be a matter of adding JS.
* A loading screen for the editor would be nice.
* Love that you handled the light/dark mode transition.

**Comment by PrateekSane** [comment]
big arrow under the cursor

**Comment by PrateekSane** [comment]
https://microsoft.github.io/monaco-editor/api/interfaces/monaco.editor.istandaloneeditorconstructionoptions.html

**Comment by krashanoff** [comment]
On account of a number of bugs stalling development, I will take over this ticket as an experimental wade into using Monaco in our browser. There's a reason that there's bugs with GitHub codespaces.

**Comment by lumisphere902** [comment]
Closing PR due to age/inactivity

**Pull request closed without merging**